### DISCUSSION TITLE
[RFC]: Shared non-graphQL schema for REST endpoints and sync webhooks

### DISCUSSION THREAD

**Discussion by lkostrowski** [saleor/saleor#17142] [RFC]
## Problem

While graphQL solves problems with strict contracts between Saleors and their clients, not everything can be achieved with graphQL

There are several cases where API contract is fragile and error-prone.

### 1. AppManifest

AppManifest is a glue between App and Saleor. Manifest is consumed by RESTful GET call.

Manifest is declared in several places:
1. Partially saved in the App model in Saleor
2. Defined in [app-sdk](https://github.com/saleor/saleor-app-sdk/blob/172de4a91ae01e5b0b376ac6993f7f7f6ee89c79/src/types.ts#L207)
3. Defined in [docs](https://docs.saleor.io/docs/3.x/developer/extending/apps/manifest)

Both SDK and docs are maintained manually and can get out of sync.

AppManifest is the target of development - it receives new fields, it requires specific validation (like the `requiresSaleorVersion` field, which is validated against semver syntax)

**There is no single source of truth for this contract**

### 2. Sync Webhooks responses

Sync webhook is a "command" that requires the app to return a specific response to complete the transaction.

The main scenarios for these webhooks are:
1. Shipping
2. Taxes
3. Payment methods and processing

Similar to AppManifest, there is no single source of truth and contract is distributed between inline validation in Saleor, app-sdk that decalres Typescript schema and manually written docs

All three are critical for e-commerce flow, and the connection must be very solid

## Proposed solutions

Main proposal is to create a schema, that will be:
1. Accessible by all consumers (Saleor, App/app-sdk, docs and others)
2. Written in common, generic format, not specific to any language
3. Easy to compile to language specific or framework specific code (Markdown docs, Python class, Typescript interface, Java interface etc)

**Pros:**
- Extracting declarative format out of source code allows to accessible lookup by everyone.
- Single source of truth
- Less human mistakes due to lack of manual implementation in several places due to automation
- Easily parsable by other systems/interop
- Inside Saleor, CODEOWNERS can be assigned to track schema changes (via PRs) across teams
- Can be used as a validation

**Cons:**
???

#### [json-schema](https://json-schema.org/)

json-schema is a popular format that can be used as a source of truth for REST payloads

It can be compiled to other syntaxes:
- [typescript](https://www.npmjs.com/package/json-schema-to-typescript)
- [python](https://pypi.org/project/json-schema-to-class/)
- [java](https://www.jsonschema2pojo.org/)
- [swagger/openAPI](https://swagger.io/docs/specification/data-models/)

There is a huge ecosystem of validators that can use json-schema as a source
https://json-schema.org/implementations.html

json-schema can work as a 1st layer of validation in Saleor controller, but also in app-sdk to verify if response is correct, before it gets send.

### Where to keep schema

Schema file can be possibly implemented in `saleor/saleor` repo or in separate, dedicated repository

#### Keep schema in Saleor repo

Use a root level directory, eg. `/schema` to store schema files.

Questions: 
- Single file or multpile files?

**Pros:**
- Keeping file in saleor/saleor makes sense, because its a "core" of the ecosystem, same as graphql.schema is stored in this repo, not outside
- Saleor can import file directly, without fetching it 
- Validation can be easily implemented. 2nd line of validation (rules that can't be represented in schema) will be in Saleor anyway. If schema is stored in the separate repository, there will be distributed validation rules
- Fetching schema in App or other clients can be done in similar way as graphql schema (`"fetch-schema": "curl https://raw.githubusercontent.com/saleor/saleor/${npm_package_saleor_schemaVersion}/saleor/graphql/schema.graphql > graphql/schema.graphql"`)

**Cons:** 
???

#### Keep schema in separate repository

Separate repository can be created just for keeping schema. 

**Pros**
- Easier to set up tooling, json-schema on every merge can generate all artifacts. This can be achievable in Saleor too.

**Cons**
- If used for validation, schema and custom validation rules will be separated
- Another repository to maintain

### When to compile

Schema as a source file can be used directly by e.g. validators, but very likely it will need to be compiled to code specific to language or framework.

For example, app-sdk that is Typescript tooling, requires typescript `interface`, likely a `*.d.ts` file.

While runtime javascript code could use schema (assuming json-schema) to validate code, it will not help with static typing, allowing great DX due to type completion. To achieve that, typescript artifact is required.

Artifacts can be generated in several ways:

#### In Saleor 

Github workflow can be prepared to build all artifacts in the same repository/folder

**Pros**
- Everything is prepared ahead of time
- Artifacts can be previewed easily if someone prefers another syntax

**Cons**
- Possible problems with syntax/versioning, for example app can use older Typescript than compiler on workflow, so artifact may not be compatible. In this scenario, app could still use source schema and compile on its own side 

#### On the client

Schema can be also fetched on any client, like app or app-sdk and compile it to only-required field there

**Props:**
- Less to maintain on the source repository
- No need to figure out how to configure compiler/versions/syntax

**Cons:**
- Worse DX / contributors need this setup
- Duplicated work on any client

**Comment by krzysztofwolski** [comment]
Would love if the OpenAPI would be chosen:
- Docs could generate automatically
- could generate types for the SDK
- a lot of tools that could be used to test both the core implementation and consumers

For the file location: not a fan of the current location of gql schema, not quite clear when to find it when browsing code the first time. `/schema` directory for gql and sync hooks would IMO improved (but require a migration period for all the old scripts to update the path)

**Comment by typeofweb** [comment]
I'm convinced we should use JSON schema. As a matter of fact, that's what we've started using in saleor-app-payment template:

https://github.com/saleor/saleor-app-payment-template/tree/canary/src/schemas

- JSON schemas
- A script that compiles them to .d.mts files (using `json-schema-to-typescript`)
- A script that compiles them to .mjs files (using `ajv/dist/standalone`)

**Comment by lkostrowski** [comment]
Dropping idea from @patrys 

Instead of building artifacts inside workflow and committing them, build them and upload to Github storage. They should be tracked per release and connected to the release, like assets here: https://github.com/saleor/saleor/releases/tag/3.12.2

All schemas should be clearly versioned

Same for generated graphql.schema - upload as artifact to the release

**Comment by korycins** [comment]
Before we introduced the subscription webhooks I was checking the possibility to add the schema for the webhooks with pydantic. Finally, we moved with the subscription webhooks as it was a lot of more benefits than adding typing to old payloads. But it sounds like a perfect thing for this RFC. Personally, I would keep this with core, as we could use pydantic models to handle the response from App, and generate the schema based on the models. 
Below you can find a sample for one webhook.
```python

from __future__ import annotations

from typing import Any, Dict, List, Optional

from enum import Enum
from pydantic import BaseModel, Field
from decimal import Decimal

class Result(str, Enum):
    CHARGE_SUCCESS = 'CHARGE_SUCCESS'
    CHARGE_FAILURE = 'CHARGE_FAILURE'


class TransactionInitializeSession(BaseModel):
    pspReference: str
    time: Optional[str]
    externalUrl: Optional[str]
    data: Dict[Any, Any] = Field(None)
    message: Optional[str] = Field(None, description="Message related to the action.")
    result: Result = Field(..., description="The final result of the transaction intialize webhook.")
    amount: Decimal = Field(..., gte=Decimal(0))



print(TransactionInitializeSession.schema_json(indent=2))
```
As a respone we will get:
```json
{
  "title": "TransactionInitializeSession",
  "type": "object",
  "properties": {
    "pspReference": {
      "title": "Pspreference",
      "type": "string"
    },
    "time": {
      "title": "Time",
      "type": "string"
    },
    "externalUrl": {
      "title": "Externalurl",
      "type": "string"
    },
    "data": {
      "title": "Data",
      "type": "object"
    },
    "message": {
      "title": "Message",
      "description": "Message related to the action.",
      "type": "string"
    },
    "result": {
      "description": "The final result of the transaction intialize webhook.",
      "allOf": [
        {
          "$ref": "#/definitions/Result"
        }
      ]
    },
    "amount": {
      "title": "Amount",
      "gte": 0,
      "type": "number"
    }
  },
  "required": [
    "pspReference",
    "result",
    "amount"
  ],
  "definitions": {
    "Result": {
      "title": "Result",
      "description": "An enumeration.",
      "enum": [
        "CHARGE_SUCCESS",
        "CHARGE_FAILURE"
      ],
      "type": "string"
    }
  }
}
```
Accodring to pydantic docs:
> The generated schemas are compliant with the specifications: [JSON Schema Core](https://json-schema.org/latest/json-schema-core.html), [JSON Schema Validation](https://json-schema.org/latest/json-schema-validation.html) and [OpenAPI](https://github.com/OAI/OpenAPI-Specification).

**Comment by andrzejewsky** [comment]
+1 to `json-schema` this is actually the most common approach and as was mentioned it's already been used in our projects.

**One point**: transformation to other languages/syntax - Please test them before any wider usage, at least for our common cases, I remember that I had some problems with it. It was years ago so probably it has significantly changed, but worth checking.

**Comment by lkostrowski** [comment]
> Before we introduced the subscription webhooks I was checking the possibility to add the schema for the webhooks with pydantic. Finally, we moved with the subscription webhooks as it was a lot of more benefits than adding typing to old payloads. But it sounds like a perfect thing for this RFC. Personally, I would keep this with core, as we could use pydantic models to handle the response from App, and generate the schema based on the models. Below you can find a sample for one webhook.
> 
> ```python
> from __future__ import annotations
> 
> from typing import Any, Dict, List, Optional
> 
> from enum import Enum
> from pydantic import BaseModel, Field
> from decimal import Decimal
> 
> class Result(str, Enum):
>     CHARGE_SUCCESS = 'CHARGE_SUCCESS'
>     CHARGE_FAILURE = 'CHARGE_FAILURE'
> 
> 
> class TransactionInitializeSession(BaseModel):
>     pspReference: str
>     time: Optional[str]
>     externalUrl: Optional[str]
>     data: Dict[Any, Any] = Field(None)
>     message: Optional[str] = Field(None, description="Message related to the action.")
>     result: Result = Field(..., description="The final result of the transaction intialize webhook.")
>     amount: Decimal = Field(..., gte=Decimal(0))
> 
> 
> 
> print(TransactionInitializeSession.schema_json(indent=2))
> ```
> 
> As a respone we will get:
> 
> ```json
> {
>   "title": "TransactionInitializeSession",
>   "type": "object",
>   "properties": {
>     "pspReference": {
>       "title": "Pspreference",
>       "type": "string"
>     },
>     "time": {
>       "title": "Time",
>       "type": "string"
>     },
>     "externalUrl": {
>       "title": "Externalurl",
>       "type": "string"
>     },
>     "data": {
>       "title": "Data",
>       "type": "object"
>     },
>     "message": {
>       "title": "Message",
>       "description": "Message related to the action.",
>       "type": "string"
>     },
>     "result": {
>       "description": "The final result of the transaction intialize webhook.",
>       "allOf": [
>         {
>           "$ref": "#/definitions/Result"
>         }
>       ]
>     },
>     "amount": {
>       "title": "Amount",
>       "gte": 0,
>       "type": "number"
>     }
>   },
>   "required": [
>     "pspReference",
>     "result",
>     "amount"
>   ],
>   "definitions": {
>     "Result": {
>       "title": "Result",
>       "description": "An enumeration.",
>       "enum": [
>         "CHARGE_SUCCESS",
>         "CHARGE_FAILURE"
>       ],
>       "type": "string"
>     }
>   }
> }
> ```
> 
> Accodring to pydantic docs:
> 
> > The generated schemas are compliant with the specifications: [JSON Schema Core](https://json-schema.org/latest/json-schema-core.html), [JSON Schema Validation](https://json-schema.org/latest/json-schema-validation.html) and [OpenAPI](https://github.com/OAI/OpenAPI-Specification).

@korycins do you suggest to build json-schema from pydantic? 

This is important decision, because json-schema is not a source anymore. Im afraid that it will cause some problems.

My proposal is to have these scenarios:
1. pydantic loads json-schema to validate (extra rules can be added on top of that, as a 2nd line of validation)
2. JS/TS code can load json-schema to tools like `ajv` to achieve the same in js/ts
3. compiles can generate json-schema -> `*.d.ts`

If we start with pydantic classes, we will have:
1. pydantic -> json-schema -> ajv
2. pydantic ->json-schema -> .d.ts

json-schema as an artifact no longer can be controlled "by hand", so if there is need to any manual improvement (syntax, description), we can be limited (unless pydantic to json-schema is super well done and covers everything)

I can't prove these problems will appear, its a general feeling that we will have 2 layers of auto-generation and it sounds dangerous

**Comment by korycins** [comment]
@lkostrowski I suggest adopting the same approach for building a JSON schema as we have for building a GraphQL schema. We already have a data class within Saleor's code that stores the details of the response that a webhook can return. Therefore, it should be feasible to convert these classes to Pydantic models and provide a command or endpoint to share them with you.

Personally, I believe it is crucial to ensure that the generated JSON schema is always related to the response models defined on the Saleor side. This way, we can guarantee that the schema is a valid contract. I am not certain if Pydantic can be customized to meet our specific requirements. Although I tested it with some old async payloads a year or two ago and had no issues, we will likely need to try it with two or three sync webhooks to confirm that it can handle all of the improvements we would like to add.

**Comment by przlada** [comment]
I agree with @korycins to start with a test. It seems that Pydantic provides ways to configure and customize JSON schema generated from its models: https://docs.pydantic.dev/usage/schema/

**Comment by lkostrowski** [comment]
@mmiszy @przlada please handover this topic from Appstore to DEX team

**Comment by typeofweb** [comment]
In @saleor/dex we don't have any competencies related to the core and/or Python. What are the exact requirements on our side once this issue is handed to us? @przlada @lkostrowski

**Comment by lkostrowski** [comment]
At this point I don't think there is anything related to core. Its purely about tooling and DX. AFAIK @przlada already created feature that builds json-schema from Python validators/model. 

To finish this, we need to take json-schema and attach additional tooling, like conversion to .d.ts, attaching it to the Saleor release artifacts etc

@przlada can pass the details

**Comment by typeofweb** [comment]
So the outstanding work will be in the CLI, correct?

**Comment by lkostrowski** [comment]
@mmiszy I think the MVP should be to attach json-schema + .d.ts to each Saleor release artifact.

Then we can connect it to app-templates and SDK to fetch it, just like we fetch schema.
If we see any CLI usages, we can add it later
